### PR TITLE
[FLINK-28623][network] Optimize the use of off heap memory by blocking and hybrid shuffle reader

### PR DESCRIPTION
**Pull request by reswqa**
## What is the purpose of the change

*Currently, each FileReader(PartitionFileReader or HsSubpartitionFileReaderImpl) will internally allocate a headerbuffer with the size of 8B. Beside, PartitionFileReader also has a 12B indexEntryBuf. Because FileReader is of subpartition granularity, If the parallelism becomes very big, and there are many slots on each TM, the memory occupation will even reach the MB level. In fact, all FileReader of the same ResultPartition read data in a single thread, so we only need to allocate a headerbuffer to a ResultPartition to optimize this phenomenon.*

## Brief change log

  - *Move `headerBuffer` and `indexEntryBuffer` from `PartitionedFileReader` to `SortMergeResultPartitionReadScheduler`*
  - *Move `headerBuffer` from `HsSubpartitionFileReaderImpl` to `HsFileDataManager`.*

## Verifying this change

This change is already covered by existing tests.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature?  no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by reswqa** [comment]
@flinkbot run azure

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartitionReadScheduler.java [hunk 1]
```diff
@@ -70,6 +71,17 @@ class SortMergeResultPartitionReadScheduler implements Runnable, BufferRecycler
      */
     private static final Duration DEFAULT_BUFFER_REQUEST_TIMEOUT = Duration.ofMinutes(5);
 
+    /** Used to read buffers from file channel. */
```
Comment (wsry) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartitionReadScheduler.java:74]: buffers -> buffer headers

**Comment by wsry** [comment]
@reswqa Thanks for the change. I left a minor comment. Please rebase master and fix the conflicts.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartitionReadScheduler.java (hunk 1 above)
Comment (reswqa) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartitionReadScheduler.java]: fixed.

**Comment by wsry** [comment]
@reswqa Thanks for the update. LGTM. Will merge after CI gives green.

**Review by wsry — APPROVED** [review]

**Pull request closed without merging**